evcc: init at 0.107.0; nixos/evcc: init; nixos/tests/evcc: init

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

https://github.com/andig/evcc

Electrovehicular charge controller software.

- [x] Testing
- [x] Module
- [x] Test

A minimal test was added to see if the service starts up with a dummy config and is reachable.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mweinelt committed:
evcc: init at 0.107.1

mweinelt committed:
nixos/evcc: init

mweinelt committed:
nixos/tests/evcc: init

andir commented on nixos/tests/evcc.nix:
```diff
@@ -0,0 +1,68 @@
+    )
+
+    with subtest("Check systemd hardening"):
+        output = machine.execute("systemd-analyze security evcc.service | grep -v '✓'")[1]
```
Ugh, this might work today but this might break whenever we upgrade systemd and they change their output format. Is there perhaps a machine readable format that you should consume instead? Also not that the severity levels might change when upgrading systemd.

I've done similar tests myself in the past but usually just during packaging to ensure I do not accidentially change something on there.

mweinelt replied:
Pretty sure that is the fallback format already, given that the normal format has emojis.

Basically the grep is there to cut down on the noise. And the test itself is there, so I can quickly see on updates if there is something I could easily improve.

Mic92 replied:
But still if systemd-analyze adds new checks it should not break the service. For the same reason we don't enable flake8 linting as upgrading it, can break perfectly fine packages.

mweinelt replied:
Why is machine.execute and issue then? It is allowed to fail if I don't misunderstand it.

Mic92 replied:
Never mind! I got that wrong.

Mic92 commented on pkgs/servers/home-automation/evcc/node-packages.nix:
```diff
@@ -0,0 +1,17910 @@
+# This file has been generated by node2nix 1.9.0. Do not edit!
```
18000 lines is a bit too much for a package I don't use but still have to download via nixpkgs.

Mic92 replied:
This would be a better solution: https://github.com/NixOS/nixpkgs/pull/128749

mweinelt replied:
Yeah, when though?

Mic92 reviewed (changes requested):
I am not in favor in having that in nixpkgs if it adds this much generated code.

SuperSandro2000 commented on pkgs/servers/home-automation/evcc/default.nix:
```diff
@@ -0,0 +1,129 @@
+  doCheck = false;
+
+  passthru.updateScript = pkgs.writeShellScript "evcc-updater" ''
+    export PATH=${lib.makeBinPath [
+      coreutils
+      curl
+      git
+      jq
+      nix-prefetch-git
+      nix-update
+      nodePackages.node2nix
+    ]}
```
This could be done with nix-shell.

roberth replied:
It could, but this cuts out the redundant evaluation that would have to be done by nix-shell.

In this case that's dwarfed by the `node2nix` invocation, so the difference is not significant. I'll assume you don't mean this as a general rule.

SuperSandro2000 commented on pkgs/servers/home-automation/evcc/default.nix:
```diff
@@ -0,0 +1,129 @@
+    nodejs
+  ];
+
+  buildFlagsArray = [
+    "-tags=release"
+  ];
```
```suggestion
  tags = [ "release" ];
```

mweinelt commented:
Updated to 0.70.

Removed the release notes, since they don't apply anymore.

> - [EVCC](https://github.com/andig/evcc) is an EV charge controller with PV integration. It supports a multitude of chargers, meters, vehicle APIs and more and ties that together with a well-tested backend and a lightweight web frontend. Available as [services.evcc](#opt-services.evcc.enable).

roberth commented:
I don't see why or how I was requested for review. The module system usage looks pretty straightforward and ok to me. The rest I can't really comment on.

SuperSandro2000 commented on pkgs/servers/home-automation/evcc/default.nix:
```diff
@@ -0,0 +1,141 @@
+
+    cd pkgs/servers/home-automation/evcc
+
+    latest=$(curl https://api.github.com/repos/evcc-io/evcc/releases/latest | jq -r '.name')
```
This should optionally take a GITHUB_TOKEN.

SuperSandro2000 commented on pkgs/servers/home-automation/evcc/default.nix:
```diff
@@ -0,0 +1,141 @@
+in
+buildGoModule rec {
+  pname = "evcc";
+  version = "";
```
How did that happen?

mweinelt replied:
Funny. Ran the update script, ran the tests, did the commit and didn't look at the changes.

SuperSandro2000 replied:
Can't comment on the line but ``src = ../../../../../../../../../tmp/tmp.TDkejGx5wj;`` in nodePackages also looks weird.

mweinelt committed:
enumer: init at 1.5.7

mweinelt commented:
Reworked to use `fetchNpmDeps` and removed the large node-packages.nix file.

cole-h commented:
@ofborg eval

mweinelt commented:
@ofborg test evcc

winterqt commented:
Marking as a draft until #200470 is merged.
